src/index.js

```diff
@@ -58,45 +58,59 @@ module.exports = class FaviconsWebpackPlugin {
-        default:
-          faviconCompilation = this.generateFaviconsWebapp(compiler, compilation);
-      }
-      
+    compiler.hooks.make.tapPromise('FaviconsWebpackPlugin', async compilation => {
+      const faviconCompilation = this.generateFavicons(compiler, compilation);
+
       // Hook into the html-webpack-plugin processing and add the html
-      tapHtml(compilation, 'FaviconsWebpackPlugin', (htmlPluginData, htmlWebpackPluginCallback) => {
-        faviconCompilation.then((tags) => {
-          if (this.options.inject(htmlPluginData.plugin)) {
-            const idx = (htmlPluginData.html + '</head>').search(/<\/head>/i);
-            htmlPluginData.html = [htmlPluginData.html.slice(0, idx), ...tags, htmlPluginData.html.slice(idx)].join('');
-          }
-          htmlWebpackPluginCallback(null, htmlPluginData);
-        }).catch(htmlWebpackPluginCallback);
-      });
+      const htmlWebpackPlugin = compiler.options.plugins
+        .map(({ constructor }) => constructor)
+        .find(({ name }) => name === 'HtmlWebpackPlugin');
+
+      if (htmlWebpackPlugin) {
+        htmlWebpackPlugin.getHooks(compilation).alterAssetTags.tapAsync('FaviconsWebpackPlugin', (htmlPluginData, htmlWebpackPluginCallback) => {
+          faviconCompilation.then((tags) => {
 
-      // Save the promise and execute the callback immediately to not block 
-      // the webpack build see the `afterCompile` FaviconsWebpackPlugin hook 
+            if (this.options.inject(htmlPluginData.plugin)) {
+              htmlPluginData.plugin.options.inject = true;
+              [].push.apply(
+                htmlPluginData.assetTags.meta,
+                tags.map(tag => parse5.parseFragment(tag).childNodes[0]).map(({ tagName, attrs }) => ({
+                  tagName,
+                  voidTag: true,
+                  attributes: attrs.reduce((obj, { name, value }) => Object.assign(obj, { [name]: value }), {}),
+                })),
+              );
+            }
+            htmlWebpackPluginCallback(null, htmlPluginData);
+          }).catch(htmlWebpackPluginCallback);
+        });
+      }
+
+      // Save the promise and execute the callback immediately to not block
+      // the webpack build see the `afterCompile` FaviconsWebpackPlugin hook
       // implementation where the promise is picked up again
       faviconCompilations.set(compilation, faviconCompilation);
-      callback();
     });
 
     // Make sure that the build waits for the favicon generation to complete
-    tap(compiler, 'afterCompile', 'FaviconsWebpackPlugin', (compilation, callback) => {
+    compiler.hooks.afterCompile.tapPromise('FaviconsWebpackPlugin', async compilation => {
       const faviconCompilation = faviconCompilations.get(compilation) || Promise.resolve();
       faviconCompilations.delete(compilation);
-      faviconCompilation.then(() => callback(), callback);
+      await faviconCompilation;
     });
+  }
 
+  generateFavicons(compiler, compilation) {
```

Comment:
Having it as its own function is way better - great choice 👍

I know I did it the same way but maybe we could also use `compilation.compiler` so this function would require only one argument?

Reply:
Sounds good, I didn't know `compiler` was in `compilation`.